# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [2]:
import pandas as pd
url1 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/titanic_train.csv'
df_1 = pd.read_csv(url1)

In [5]:
import pandas as pd
url2 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/titanic_train.csv'
df_2 = pd.read_csv(url2)

In [6]:
import pandas as pd
url3 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/titanic_train.csv'
df_3 = pd.read_csv(url3)

In [7]:
df_1.sample(10)

PassengerId  Survived  Pclass  \
887          888         1       1   
654          655         0       3   
135          136         0       2   
594          595         0       2   
826          827         0       3   
667          668         0       3   
212          213         0       3   
292          293         0       2   
18            19         0       3   
368          369         1       3   

                                                  Name     Sex   Age  SibSp  \
887                       Graham, Miss. Margaret Edith  female  19.0      0   
654                       Hegarty, Miss. Hanora "Nora"  female  18.0      0   
135                                 Richard, Mr. Emile    male  23.0      0   
594                            Chapman, Mr. John Henry    male  37.0      1   
826                                       Lam, Mr. Len    male   NaN      0   
667                         Rommetvedt, Mr. Knud Paust    male   NaN      0   
212                             Perkin, Mr. John Henry    male  22.0      0   
292                             Levy, Mr. Rene Jacques    male  36.0      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
368                                Jermyn, Miss. Annie  female   NaN      0   

     Parch         Ticket     Fare Cabin Embarked  
887      0         112053  30.0000   B42        S  
654      0         365226   6.7500   NaN        Q  
135      0  SC/PARIS 2133  15.0458   NaN        C  
594      0    SC/AH 29037  26.0000   NaN        S  
826      0           1601  56.4958   NaN        S  
667      0         312993   7.7750   NaN        S  
212      0      A/5 21174   7.2500   NaN        S  
292      0  SC/Paris 2163  12.8750     D        C  
18       0         345763  18.0000   NaN        S  
368      0          14313   7.7500   NaN        Q

In [8]:
df_2.sample(10)

PassengerId  Survived  Pclass  \
847          848         0       3   
265          266         0       2   
858          859         1       3   
626          627         0       2   
559          560         1       3   
82            83         1       3   
821          822         1       3   
116          117         0       3   
755          756         1       2   
383          384         1       1   

                                                  Name     Sex    Age  SibSp  \
847                                 Markoff, Mr. Marin    male  35.00      0   
265                                  Reeves, Mr. David    male  36.00      0   
858              Baclini, Mrs. Solomon (Latifa Qurban)  female  24.00      0   
626                     Kirkland, Rev. Charles Leonard    male  57.00      0   
559       de Messemaeker, Mrs. Guillaume Joseph (Emma)  female  36.00      1   
82                      McDermott, Miss. Brigdet Delia  female    NaN      0   
821                                  Lulic, Mr. Nikola    male  27.00      0   
116                               Connors, Mr. Patrick    male  70.50      0   
755                          Hamalainen, Master. Viljo    male   0.67      1   
383  Holverson, Mrs. Alexander Oskar (Mary Aline To...  female  35.00      1   

     Parch      Ticket     Fare Cabin Embarked  
847      0      349213   7.8958   NaN        C  
265      0  C.A. 17248  10.5000   NaN        S  
858      3        2666  19.2583   NaN        C  
626      0      219533  12.3500   NaN        Q  
559      0      345572  17.4000   NaN        S  
82       0      330932   7.7875   NaN        Q  
821      0      315098   8.6625   NaN        S  
116      0      370369   7.7500   NaN        Q  
755      1      250649  14.5000   NaN        S  
383      0      113789  52.0000   NaN        S

In [9]:
df_3.sample(10)

PassengerId  Survived  Pclass  \
849          850         1       1   
590          591         0       3   
167          168         0       3   
152          153         0       3   
262          263         0       1   
83            84         0       1   
671          672         0       1   
349          350         0       3   
288          289         1       2   
240          241         0       3   

                                                Name     Sex   Age  SibSp  \
849     Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female   NaN      1   
590                             Rintamaki, Mr. Matti    male  35.0      0   
167  Skoog, Mrs. William (Anna Bernhardina Karlsson)  female  45.0      1   
152                                 Meo, Mr. Alfonzo    male  55.5      0   
262                                Taussig, Mr. Emil    male  52.0      1   
83                           Carrau, Mr. Francisco M    male  28.0      0   
671                           Davidson, Mr. Thornton    male  31.0      1   
349                                 Dimic, Mr. Jovan    male  42.0      0   
288                             Hosono, Mr. Masabumi    male  42.0      0   
240                            Zabour, Miss. Thamine  female   NaN      1   

     Parch             Ticket     Fare Cabin Embarked  
849      0              17453  89.1042   C92        C  
590      0  STON/O 2. 3101273   7.1250   NaN        S  
167      4             347088  27.9000   NaN        S  
152      0         A.5. 11206   8.0500   NaN        S  
262      1             110413  79.6500   E67        S  
83       0             113059  47.1000   NaN        S  
671      0         F.C. 12750  52.0000   B71        S  
349      0             315088   8.6625   NaN        S  
288      0             237798  13.0000   NaN        S  
240      0               2665  14.4542   NaN        C

- Elección de concat porque los 3 df tienen la mismas estructura. Lo hago por filas (axis=0)

In [12]:
df_final = pd.concat([df_1, df_2, df_3], axis=0)

- Revisión inicial del df.

In [23]:
df_final.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [24]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2673 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  2673 non-null   int64  
 1   Survived     2673 non-null   int64  
 2   Pclass       2673 non-null   int64  
 3   Name         2673 non-null   object 
 4   Sex          2673 non-null   object 
 5   Age          2142 non-null   float64
 6   SibSp        2673 non-null   int64  
 7   Parch        2673 non-null   int64  
 8   Ticket       2673 non-null   object 
 9   Fare         2673 non-null   float64
 10  Cabin        612 non-null    object 
 11  Embarked     2667 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 271.5+ KB


In [25]:
df_final.describe()

PassengerId     Survived       Pclass          Age        SibSp  \
count  2673.000000  2673.000000  2673.000000  2142.000000  2673.000000   
mean    446.000000     0.383838     2.308642    29.699118     0.523008   
std     257.257509     0.486410     0.835758    14.519711     1.102331   
min       1.000000     0.000000     1.000000     0.420000     0.000000   
25%     223.000000     0.000000     2.000000    20.000000     0.000000   
50%     446.000000     0.000000     3.000000    28.000000     0.000000   
75%     669.000000     1.000000     3.000000    38.000000     1.000000   
max     891.000000     1.000000     3.000000    80.000000     8.000000   

             Parch         Fare  
count  2673.000000  2673.000000  
mean      0.381594    32.204208  
std       0.805755    49.674827  
min       0.000000     0.000000  
25%       0.000000     7.895800  
50%       0.000000    14.454200  
75%       0.000000    31.000000  
max       6.000000   512.329200

- Renombrar columnas.

In [26]:
df_final.columns = df_final.columns.str.lower()

-------------

In [28]:
df_final

passengerid  survived  pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  name     sex   age  sibsp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     parch            ticket     fare cabin embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[2673 rows x 12 columns]

In [27]:
df_final.dtypes

passengerid      int64
survived         int64
pclass           int64
name            object
sex             object
age            float64
sibsp            int64
parch            int64
ticket          object
fare           float64
cabin           object
embarked        object
dtype: object

In [29]:
df_final.isnull().sum()

passengerid       0
survived          0
pclass            0
name              0
sex               0
age             531
sibsp             0
parch             0
ticket            0
fare              0
cabin          2061
embarked          6
dtype: int64

- Cabin tiene 2061 nulos de 2673 filas, elimino la columna

In [30]:
df_final.drop(['cabin'], axis=1, inplace=True)

In [31]:
df_final

passengerid  survived  pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  name     sex   age  sibsp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     parch            ticket     fare embarked  
0        0         A/5 21171   7.2500        S  
1        0          PC 17599  71.2833        C  
2        0  STON/O2. 3101282   7.9250        S  
3        0            113803  53.1000        S  
4        0            373450   8.0500        S  
..     ...               ...      ...      ...  
886      0            211536  13.0000        S  
887      0            112053  30.0000        S  
888      2        W./C. 6607  23.4500        S  
889      0            111369  30.0000        C  
890      0            370376   7.7500        Q  

[2673 rows x 11 columns]

--------------------------------

- age tiene 531 nulos de 2673, opto por calcular la mediana de edad y asignarselo a esos valores para no perder información de la tabla de esas columnas

- Le asigno el tipo correcto.

In [32]:
mediana = df_final['age'].median()

In [34]:
df_final['age'] = df_final['age'].fillna(mediana)

In [35]:
df_final['age'] = df_final['age'].astype(int)

---------------------------

- Le asigno la moda a los 6 nulos de embarked al ser str

In [37]:
moda = df_final['embarked'].mode()[0]  # Devuelve una serie, cojo el primer valor

In [38]:
df_final['embarked'] = df_final['embarked'].fillna(moda)

-----------------------

In [39]:
df_final.isnull().sum()

passengerid    0
survived       0
pclass         0
name           0
sex            0
age            0
sibsp          0
parch          0
ticket         0
fare           0
embarked       0
dtype: int64

In [45]:
df_final.sample(20)

passengerid  survived  pclass  \
359          360         1       3   
669          670         1       1   
779          780         1       1   
289          290         1       3   
41            42         0       2   
849          850         1       1   
549          550         1       2   
100          101         0       3   
712          713         1       1   
248          249         1       1   
546          547         1       2   
783          784         0       3   
140          141         0       3   
187          188         1       1   
343          344         0       2   
394          395         1       3   
536          537         0       1   
580          581         1       2   
409          410         0       3   
358          359         1       3   

                                                  name     sex  age  sibsp  \
359                  Mockler, Miss. Helen Mary "Ellie"  female   28      0   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female   28      1   
779  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...  female   43      0   
289                               Connolly, Miss. Kate  female   22      0   
41   Turpin, Mrs. William John Robert (Dorothy Ann ...  female   27      1   
849       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female   28      1   
549                     Davies, Master. John Morgan Jr    male    8      1   
100                            Petranec, Miss. Matilda  female   28      0   
712                           Taylor, Mr. Elmer Zebley    male   48      1   
248                      Beckwith, Mr. Richard Leonard    male   37      1   
546                  Beane, Mrs. Edward (Ethel Clarke)  female   19      1   
783                             Johnston, Mr. Andrew G    male   28      1   
140                      Boulos, Mrs. Joseph (Sultana)  female   28      0   
187      Romaine, Mr. Charles Hallace ("Mr C Rolmane")    male   45      0   
343         Sedgwick, Mr. Charles Frederick Waddington    male   25      0   
394  Sandstrom, Mrs. Hjalmar (Agnes Charlotta Bengt...  female   24      0   
536                  Butt, Major. Archibald Willingham    male   45      0   
580                        Christy, Miss. Julie Rachel  female   25      1   
409                                 Lefebre, Miss. Ida  female   28      3   
358                               McGovern, Miss. Mary  female   28      0   

     parch      ticket      fare embarked  
359      0      330980    7.8792        Q  
669      0       19996   52.0000        S  
779      1       24160  211.3375        S  
289      0      370373    7.7500        Q  
41       0       11668   21.0000        S  
849      0       17453   89.1042        C  
549      1  C.A. 33112   36.7500        S  
100      0      349245    7.8958        S  
712      0       19996   52.0000        S  
248      1       11751   52.5542        S  
546      0        2908   26.0000        S  
783      2  W./C. 6607   23.4500        S  
140      2        2678   15.2458        C  
187      0      111428   26.5500        S  
343      0      244361   13.0000        S  
394      2     PP 9549   16.7000        S  
536      0      113050   26.5500        S  
580      1      237789   30.0000        S  
409      1        4133   25.4667        S  
358      0      330931    7.8792        Q

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [50]:
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv'
df_insurance = pd.read_csv(url)

In [51]:
df_insurance.sample(20)

unnamed:_0 customer       state  customer_lifetime_value response  \
1247         1247  HP55853  California             22332.439460      Yes   
3609         3609  FD11364      Oregon              6768.997027       No   
1253         1253  BB10681  California              4140.648654       No   
3505         3505  QF21628  California              3095.627124       No   
2565         2565  KR50062      Oregon              2754.456760       No   
4833         4833  QW59726  California              2944.780529       No   
8268         8268  KX86150  California              7188.292633       No   
3192         3192  AA34092  California             28799.954150       No   
8155         8155  HX72126      Nevada             21112.536240       No   
282           282  DO18118      Oregon             10996.713360       No   
10233       10233  KC11055      Nevada             16936.271470       No   
10057       10057  BC66536     Arizona              6883.686618       No   
4388         4388  BG80246     Arizona              5570.177550       No   
2198         2198  ZN45349      Oregon              4815.020779       No   
3146         3146  AO74776  California              9427.684903       No   
10833       10833  OQ99704  Washington              3981.775889       No   
6833         6833  CK35310      Nevada              2839.569680       No   
1586         1586  BJ18904     Arizona              3964.534974       No   
10203       10203  US63915     Arizona              6218.662593       No   
2861         2861  SO54368      Oregon              4686.469674      Yes   

       coverage             education effective_to_date employmentstatus  \
1247   Extended               College        2011-01-11         Employed   
3609   Extended              Bachelor        2011-01-31         Employed   
1253   Extended               College        2011-01-17       Unemployed   
3505   Extended  High School or Below        2011-01-15         Employed   
2565      Basic               College        2011-02-11         Employed   
4833      Basic                Master        2011-01-12         Employed   
8268   Extended  High School or Below        2011-01-29    Medical Leave   
3192   Extended               College        2011-02-11         Employed   
8155   Extended              Bachelor        2011-02-22       Unemployed   
282       Basic  High School or Below        2011-01-02       Unemployed   
10233  Extended              Bachelor        2011-01-24         Employed   
10057  Extended  High School or Below        2011-02-26         Employed   
4388      Basic               College        2011-01-11         Disabled   
2198      Basic               College        2011-01-16       Unemployed   
3146      Basic  High School or Below        2011-01-21         Employed   
10833     Basic              Bachelor        2011-02-11         Employed   
6833      Basic  High School or Below        2011-01-02         Employed   
1586      Basic               College        2011-02-22    Medical Leave   
10203   Premium  High School or Below        2011-02-04         Employed   
2861      Basic               College        2011-02-12         Employed   

      gender  ...  number_of_policies     policy_type        policy  \
1247       M  ...                   2  Corporate Auto  Corporate L3   
3609       F  ...                   3    Special Auto    Special L3   
1253       F  ...                   1  Corporate Auto  Corporate L3   
3505       M  ...                   1   Personal Auto   Personal L3   
2565       M  ...                   1   Personal Auto   Personal L3   
4833       F  ...                   1   Personal Auto   Personal L2   
8268       M  ...                   8   Personal Auto   Personal L3   
3192       M  ...                   2   Personal Auto   Personal L3   
8155       F  ...                   2   Personal Auto   Personal L3   
282        F  ...                   2   Personal Auto   Personal L2   
10233      F  ...               

In [53]:
pd.set_option('display.max_columns', None)

In [55]:
df_insurance.sample(10)

unnamed:_0 customer       state  customer_lifetime_value response  \
4498        4498  JL25091  California              4599.759229       No   
9790        9790  OF38237  California              4878.829713       No   
1669        1669  HR64684      Oregon              8689.822524       No   
6600        6600  KG63751  Washington              6751.437153       No   
5111        5111  HK40573  California             14382.564630       No   
5108        5108  WG51330  California              5635.294733       No   
5417        5417  ZW33889  California              2477.708203       No   
665          665  SM42781     Arizona             14040.964760       No   
3724        3724  CQ72935     Arizona              4974.801539      Yes   
8863        8863  ED91989     Arizona              7999.251628       No   

      coverage             education effective_to_date employmentstatus  \
4498   Premium              Bachelor        2011-02-27         Employed   
9790     Basic                Doctor        2011-02-28         Disabled   
1669  Extended                Master        2011-02-08       Unemployed   
6600  Extended  High School or Below        2011-02-23       Unemployed   
5111  Extended               College        2011-01-08         Employed   
5108     Basic               College        2011-01-25         Employed   
5417     Basic              Bachelor        2011-02-14         Employed   
665      Basic              Bachelor        2011-02-22       Unemployed   
3724     Basic              Bachelor        2011-01-18         Employed   
8863     Basic               College        2011-02-01       Unemployed   

     gender  income location_code marital_status  monthly_premium_auto  \
4498      M   79284         Rural         Single                   117   
9790      F   29950      Suburban        Married                    61   
1669      M       0      Suburban       Divorced                    99   
6600      M       0      Suburban        Married                    95   
5111      M   52174         Urban        Married                   121   
5108      M   59325         Urban        Married                    71   
5417      F   44280         Rural        Married                    62   
665       F       0         Urban        Married                   195   
3724      M   75644      Suburban       Divorced                    65   
8863      M       0      Suburban        Married                   107   

      months_since_last_claim  months_since_policy_inception  \
4498                      9.0                             59   
9790                      0.0                             76   
1669                     25.0                             57   
6600                     14.0                             78   
5111                     22.0                             25   
5108                     13.0                             75   
5417                     25.0                              3   
665                       9.0                             33   
3724                      5.0                             68   
8863                     13.0                             17   

      number_of_open_complaints  number_of_policies     policy_type  \
4498                        1.0                   1    Special Auto   
9790                        0.0                   4   Personal Auto   
1669                        5.0                   2   Personal Auto   
6600                        0.0                   9   Personal Auto   
5111                        0.0                   2   Personal Auto   
5108                        0.0                   3  Corporate Auto   
5417                        0.0                   1  Corporate Auto   
665                         1.0                   5   Personal Auto   
3724                        3.0                   9   Personal Auto   
8863                        0.0                   5   Personal Auto   

            policy renew_offer_type sales_channel  t

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

- ganancias totales = ingresos mensuales * nº de meses contratados - gastos
- Calcular el min() de nº de meses contratados para hacer una comparación entre todas las filas de forma igualitaria.

In [56]:
columnas_seleccionadas = ['monthly_premium_auto', 'months_since_policy_inception', 'sales_channel', 'total_claim_amount']
df_c2 = df_insurance[columnas_seleccionadas]

In [57]:
df_c2.head()

monthly_premium_auto  months_since_policy_inception sales_channel  \
0                    61                             52         Agent   
1                    64                             26   Call Center   
2                   100                             31   Call Center   
3                    97                              3        Branch   
4                   117                             31        Branch   

   total_claim_amount  
0          292.800000  
1          744.924331  
2          480.000000  
3          484.013411  
4          707.925645

In [60]:
valor_minimo = df_c2['months_since_policy_inception'].min()
valor_minimo

0

In [62]:
frecuencia_valores = df_c2['months_since_policy_inception'].value_counts()
frecuencia_valores

months_since_policy_inception
59    168
61    151
50    151
10    138
3     136
     ... 
16     81
96     80
82     71
98     64
97     59
Name: count, Length: 100, dtype: int64

In [64]:
df_c2.isnull().sum()

monthly_premium_auto             0
months_since_policy_inception    0
sales_channel                    0
total_claim_amount               0
dtype: int64

- Quiero obtener la duración del percentil 25 para calcular mi estrategia de marketing basada en los datos de contratación más recientes.

In [66]:
percentil_25 = df_c2['months_since_policy_inception'].quantile(0.25)
percentil_25

24.0

- Filtro los clientes que contrataron hace 24 meses o antes.

In [67]:
df_c2 = df_c2[df_c2['months_since_policy_inception'] <= percentil_25]

In [69]:
df_c2.sample(10)

monthly_premium_auto  months_since_policy_inception sales_channel  \
6202                    71                             24           Web   
2791                   184                             10         Agent   
7241                    62                              7        Branch   
4252                    67                             18        Branch   
2671                   106                             17        Branch   
5156                   109                             23         Agent   
1547                    72                             19        Branch   
5704                    67                             13         Agent   
7518                    76                             22           Web   
5588                    62                              3        Branch   

      total_claim_amount  
6202          340.800000  
2791          883.200000  
7241          128.969729  
4252          309.076476  
2671          273.250112  
5156           25.292113  
1547          515.122954  
5704          126.634634  
7518          364.800000  
5588          266.932542

In [70]:
df_c2.shape

(2748, 4)

- Tabla resumen con pivot.

- Ganancia para cada fila:

In [71]:
df_c2['ganancia'] = df_c2['monthly_premium_auto'] * df_c2['months_since_policy_inception'] - df_c2['total_claim_amount']

- Tabla dinámica agrupada por 'sales_channel' y sumando la columna 'ganancia'

In [73]:
df_resumen = pd.pivot_table(df_c2, values='ganancia', index='sales_channel', aggfunc='sum')

In [74]:
df_resumen = df_resumen.round(2)

In [75]:
df_resumen

ganancia
sales_channel           
Agent          752333.91
Branch         499397.51
Call Center    382689.54
Web            279775.49

------------------------

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [76]:
df_acl = pd.pivot_table(df_insurance, values='customer_lifetime_value', index='gender', columns='education', aggfunc='mean')

In [77]:
df_acl = df_acl.round(2)

In [78]:
df_acl

education  Bachelor  College   Doctor  High School or Below   Master
gender                                                              
F           7874.27  7748.82  7328.51               8675.22  8157.05
M           7703.60  8052.46  7415.33               8149.69  8168.83

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [4]:
# Your code goes here